### PR TITLE
Rework cgo merkletree

### LINKED ISSUE #1270: Go tests for merkletree fail
$ go test ./...
ok  	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go	0.862s
ok  	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/asn1	0.050s
ok  	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/client	15.095s
ok  	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/fixchain	0.242s
?   	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/fixchain/main	[no test files]
ok  	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/fixchain/ratelimiter	27.779s
ok  	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/gossip	5.962s
?   	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/gossip/main	[no test files]
--- FAIL: TestAddLeaf (0.00s)
panic: runtime error: cgo argument has Go pointer to Go pointer [recovered]
	panic: runtime error: cgo argument has Go pointer to Go pointer

goroutine 6 [running]:
panic(0x57cb80, 0xc82000a620)
	/usr/lib/google-golang/src/runtime/panic.go:483 +0x3ef
testing.tRunner.func1(0xc82001c120)
	/usr/lib/google-golang/src/testing/testing.go:467 +0x192
panic(0x57cb80, 0xc82000a620)
	/usr/lib/google-golang/src/runtime/panic.go:443 +0x4f2
_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/merkletree.(*CPPMerkleTree).CurrentRoot(0xc8200103e0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/merkletree/merkle_tree.go:55 +0x15e
_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/merkletree.TestAddLeaf(0xc82001c120)
	/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/merkletree/merkle_tree_test.go:79 +0x481
testing.tRunner(0xc82001c120, 0x87eed8)
	/usr/lib/google-golang/src/testing/testing.go:473 +0x98
created by testing.RunTests
	/usr/lib/google-golang/src/testing/testing.go:587 +0x892
FAIL	_/usr/local/google/home/hadfieldp/ct/certificate-transparency/go/merkletree	0.026s

### PR DESCRIPTION
**Pull request by AlCutter**
Fixes #1270

### CODE REVIEW

File: go/merkletree/merkle_tree_go.cc [hunk 1]
```diff
@@ -47,109 +42,100 @@ size_t LeafCount(TREE tree) {
...
 }
 
-bool LeafHash(TREE tree, BYTE_SLICE out, size_t leaf) {
-  GoSlice* slice(BS(out));
+size_t LeafHash(TREE tree, size_t leaf, void* buf, size_t buf_len) {
   const MerkleTree* t(MT(tree));
   const size_t nodesize(t->NodeSize());
-  if (slice->data == NULL || slice->cap < nodesize) {
+  if (buf == NULL || buf_len < nodesize) {
     return false;
```
Comment (Martin2112) [inline go/merkletree/merkle_tree_go.cc]: returns false but signature has changed to size_t.

File: go/merkletree/merkle_tree_go.cc [hunk 2]
```diff
@@ -47,109 +42,100 @@ size_t LeafCount(TREE tree) {
...
 }
 
-bool CurrentRoot(TREE tree, BYTE_SLICE out) {
-  GoSlice* slice(BS(out));
+size_t CurrentRoot(TREE tree, void* buf, size_t buf_len) {
   MerkleTree* t(MT(tree));
   const size_t nodesize(t->NodeSize());
-  if (slice->data == NULL || slice->len != nodesize) {
+  if (buf == NULL || buf_len < nodesize) {
     return false;
```
Comment (Martin2112) [inline go/merkletree/merkle_tree_go.cc]: also returns false but signature has changed to size_t.

File: go/merkletree/merkle_tree_go.cc [hunk 3]
```diff
@@ -47,109 +42,100 @@ size_t LeafCount(TREE tree) {
...
+  if (buf == nullptr || buf_len < nodesize) {
+    return 0;
   }
   const std::string& hash = t->RootAtSnapshot(snapshot);
   assert(nodesize == hash.size());
-  memcpy(slice->data, hash.data(), nodesize);
-  slice->len = nodesize;
+  memcpy(buf, hash.data(), nodesize);
+  return nodesize;
   return true;
```
Comment (Martin2112) [inline go/merkletree/merkle_tree_go.cc]: two returns

Comment (AlCutter) [inline go/merkletree/merkle_tree_go.cc]: Done

File: go/merkletree/merkle_tree_go.cc (hunk 2 above)
Comment (AlCutter) [inline go/merkletree/merkle_tree_go.cc]: done

File: go/merkletree/merkle_tree_go.cc (hunk 1 above)
Comment (AlCutter) [inline go/merkletree/merkle_tree_go.cc]: done

**Pull request closed without merging**